test/index.test.ts

```diff
@@ -613,385 +613,507 @@ describe('HipThrusTS', () => {
 
         const triple = HTPipe(left, midNotCovered, rightFullyCovered);
       });
-      it('attaches properly typed data from left and right sync data attacher', async () => {
-        const aPassedIn = 'some string';
-        const bPassedIn = 4;
-        const cReturned = 6;
-        const left = {
-          attachData(context: { a: string }) {
-            expect(context.a).to.be.equal(aPassedIn);
-            return { b: bPassedIn };
-          },
-        };
 
-        const rightFullyCovered = {
-          attachData(context: { b: number }) {
-            expect(context.b).to.be.equal(bPassedIn);
-            return { c: cReturned };
-          },
+      function basicPipedCaseTest<
+        TStage extends AllStageKeys,
+        TReturnSync extends 'sync' | 'async'
+      >(stage: TStage, returnSync: TReturnSync) {
+        const testConstants = {
+          aPassedIn: 'some string',
+          bPassedIn: 4,
+          cReturned: 6,
         };
 
-        const pipedAtoBC = HTPipe(left, rightFullyCovered);
-
-        await HTPipeTest(
-          pipedAtoBC,
-          'attachData',
-          { a: aPassedIn },
-          { b: bPassedIn, c: cReturned },
-          true
-        );
-      });
-      it('attaches right sync output instead of left sync output if right output sync transform left output type', async () => {
-        const aPassedIn = 5;
-        const bPassedIn = 'some string';
-        const bReturned = 4;
-
-        const left = {
-          attachData(context: { a: number }) {
-            expect(context.a).to.be.equal(aPassedIn);
-            return { b: bPassedIn };
-          },
+        const leftProjector = (htCtx: { a: string }) => {
+          expect(htCtx.a).to.be.equal(testConstants.aPassedIn);
+          return returnSync === 'sync'
+            ? { b: testConstants.bPassedIn }
+            : Promise.resolve({ b: testConstants.bPassedIn });
         };
```

Comment:
hmm I'm a little surprised this works above... I would think that this would make the return type of leftProjector {blah} | Promise<{blah}> ... I see it's generic but I've seen typescript not be smart enough in cases like these.

I would have thought you'd need to make it different so that the individual invocation of basicPipedCaseTest has the specifics of whether leftProjector and rightProjector are sync or async.

Can you please hover over the "left" of "basicPipedCaseTest(lifecycleStage, stageSync).left" on line 913 (new) and make sure it KNOWS it's sync?

- [ ] MAYBE: typing of left/right async vs sync

Reply:
done with wrapProjectorReturnWithPromise function